MINOR: Gradle build optimizations

jthurne opened the pull request:
Applies various build optimizations to take full advantage of Gradle's incremental build feature, and Gradle's local build caching feature. The optimizations will help to reduce build times for developers when running builds locally.

Specific optimizations:

  - **Enable the local build cache**
     The local build cache avoids re-running tasks whose inputs have not changed, even if the build output has been cleaned or if the build is executed in a different directory.

  - **Instruct Gradle to ignore version files when doing up-to-date checks**
     This can save a lot of developer time when working locally and creating a lot of commits. In particular, tests won't be re-executed simply because the commit ID changes (unless one of the other test inputs changes).

  - **Update createVersionFile tasks support incremental builds**
     The version files are no longer recreated unless one of their inputs changes (the commit ID or the version property).

  - **Set the root project name in settings.gradle**
     The root project name is used to name the build as a whole. It is used in reports and other types of artifacts. If a root project name is not set, then the containing directory's name is used as the root project name.

     It is a best-practice to explicitly set the root project name. Otherwise, the root project name could change if the directory the project belongs to is different from the desired name (for example, of someone checks out the repository into a directory named differently from the default).

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

jthurne committed:
Update createVersionFile tasks to support incremental building

...by declaring their inputs

jthurne committed:
Enable local build cache

The local build cache avoids re-running tasks whose inputs have not changed,
even if the build output has been cleaned or if the build is executed in a
different directory.

jthurne committed:
Instruct Gradle to ignore version files when doing up-to-date checks

This can save a lot of time when working locally and creating a lot of commits
In particular, tests won't be re-executed simply because the git commit SHA
changes (unless one of the other test inputs changes).

jthurne committed:
Make the createVersionFile tasks cacheable

The createVersionFileTasks already declare their inputs and outputs, and aren't
disk-heavy, so we can cache them to gain a tiny bit more build efficiency.

jthurne committed:
Set the root project name in settings.gradle

The root project name is used to name the build as a whole. It is used in
reports and other types of artifacts. If a root project name is not set, then
the containing directory's name is used as the root project name.

It's a best-practice to explicitly set the root project name. Otherwise, the
root project name could change if the directory the project belongs to is
different from the desired name (for example, of someone checks out the
repository into a directory named differently from the default).

jthurne commented:
@ijuma it looks like you review a lot of these changes.

jthurne commented:
Full disclosure: I'm an engineer that works for Gradle (on [Gradle Enterprise](https://gradle.com/), and I found these optimizations while using the project to test out Gradle Enterprise features.

Also, FWIW, I was able to reduce overall test execution time by more than half by using [Gradle Enterprise's support for Test Distribution](https://gradle.com/blog/gradle-enterprise-unveils-test-distribution/). This might be something worth considering at some point in the future.

jthurne commented:
One last comment: these changes should be compatible with Gradle 7.

ijuma commented:
Thanks for the PR! Looks like the build failed:

> [2021-04-06T20:21:12.201Z] * What went wrong:
> 
> [2021-04-06T20:21:12.201Z] A problem occurred evaluating settings 'Kafka_kafka-pr_PR-10490'.
> 
> [2021-04-06T20:21:12.201Z] > No signature of method: org.gradle.initialization.DefaultProjectDescriptor.name() is applicable for argument types: (String) values: [kafka]
> 
> [2021-04-06T20:21:12.201Z]   Possible solutions: any(), wait(), path(), dump(), setName(java.lang.String), any(groovy.lang.Closure)

Also, have you measured a performance improvement without Gradle Enterprise?

runningcode commented on settings.gradle:
```diff
@@ -49,3 +51,9 @@ include 'clients',
     'streams:upgrade-system-tests-27',
     'tools'
+
+buildCache {
```
This `buildCache` block is optional and can be removed.`org.gradle.caching=true` sets up the local build cache by default with the remote build cache disabled.

jthurne replied:
I like the idea of keeping this in settings.gradle since it's more "discoverable", and thus easier to see what is and isn't enabled caching-wise.  Settings in gradle.properties tend to be forgotten about.

But to keep the setting in one place, I'll remove it from gradle.properties.

ijuma replied:
It seemed better to have this in `gradle.properties` since we define the parallelism there too. Generally, no-one ever looks at `settings.gradle` as far as I can tell (i.e. it's not more discoverable).

runningcode commented on build.gradle:
```diff
@@ -1148,8 +1155,10 @@ project(':clients') {
+    inputs.property('version', version)
     outputs.file receiptFile
-    outputs.upToDateWhen { false }
+    outputs.cacheIf { true }
```
Since this file is rather small and easy to create. It might make sense to _not_ enable the build cache for this task since the ziping and unzipping of this cache entry might take longer than it would take to write this file directly.

jthurne replied:
Yeah, that's true. In this case, the cost to zip up this small file is also tiny, but either way, there's really no benefit to caching on these tasks since they run so fast anyway.  I'll go ahead and remove this.

runningcode reviewed (approved):
Looks good! Just some nits.

jthurne committed:
Fix setting the root project name

jthurne committed:
Remove redundant setting to enable build caching

Local build caching is already enabled in settings.gradle.

jthurne committed:
Disable caching the createVersionFile tasks

The tasks run very quickly and produce a very small text file. Thus, there is
little benefit (possibily a negative impact) to caching the task output.

jthurne commented:
> Thanks for the PR! Looks like the build failed:

Whoops! That was silly of me. Somehow I accidentally deleted an important equals sign when getting these commits ready for submission. Should be fixed now (tested locally this time).

> Also, have you measured a performance improvement without Gradle Enterprise?

For locally executed builds, the build caching can really help avoid re-running tasks with inputs that haven't changed since the last time they were executed. This can save a lot of time when working locally.  You'll observe a noticeable performance gain when running clean builds. Additionally, tests aren't executed anymore if all you do is make a commit that changes something not directly related to the tests themselves, like build logic (although if you change the test tasks, then they'll of course re-run).

On the CI-server side, these changes have little to no impact at all, unfortunately.

This is already a pretty mature build and the optimizations in this PR are relatively minor. By far, the vast majority of the build time is taken up executing tests. In particular, the tests in the `core` module tend to take a long time to execute, even with parallel test execution enabled.

runningcode commented:
> On the CI-server side, these changes have little to no impact at all, unfortunately.

The CI build improvements will come from using the Gradle Enteprise remote build cache on CI builds to re-use outputs from previous builds. There is no remote build cache configured we cannot see these improvement.

jthurne committed:
Reenable the local build cache

The setting in gradle.properties acts as a master switch for the build cache,
and it defaults to false. It does not matter if the build cache is enabled in
settings.gradle, of this property if false then the cache will not be used.

jthurne committed:
Removes explicit build cache settings from settings.gradle

The settings mimic the default values, and it's the org.gradle.caching property
in gradle.properties that truly controls whether or not build caching is
enabled.

jthurne committed:
Merge remote-tracking branch 'origin/trunk' into build-optimizations

* origin/trunk:
  MINOR: move NoOpSnapshotWriter to main (#10496)
  KAFKA-10769 Remove JoinGroupRequest#containsValidPattern as it is dup… (#9851)
  KAFKA-12384: stabilize ListOffsetsRequestTest#testResponseIncludesLeaderEpoch (#10389)
  KAFKA-5146: remove Connect dependency from Streams module (#10131)
  KAFKA-12602: Fix LICENSE file (#10474)
  MINOR: Enable scala/java joint compilation consistently for `core` module (#10485)

ijuma commented:
> For locally executed builds, the build caching can really help avoid re-running tasks with inputs that haven't changed since the last time they were executed. This can save a lot of time when working locally. You'll observe a noticeable performance gain when running clean builds. Additionally, tests aren't executed anymore if all you do is make a commit that changes something not directly related to the tests themselves, like build logic (although if you change the test tasks, then they'll of course re-run).

Does the `clean` task delete the local caches or not? If so, how would one force that in case of a bug? It would be good to update the `Readme.md` with this info.

jthurne commented:
> Does the clean task delete the local caches or not?

It does not delete the local build cache.

> If not, how would one force cache deletion in case of a bug?

It's best to address bugs like those by finding the offending task and fixing it to properly declare it's inputs and outputs (or if the issue is with a third-party plugin, to file a bug report against the plugin).  

But in a pinch, you can pass `--rerun-tasks` to instruct Gradle to skip the cache and up-to-date checks. If a cache entry has been corrupted, this will also tell Gradle to overwrite the cache entry with the new task output.

However, these types of issues are quite rare. It is also best to avoid training users to use `--rerun-tasks` (as it eliminates all of the incremental build and build cache benefits).  For these reasons, I'm somewhat hesitant to update the README.

jthurne committed:
Merge remote-tracking branch 'origin/trunk' into build-optimizations

* origin/trunk: (27 commits)
  MINOR: Remove some unnecessary cyclomatic complexity suppressions (#10488)
  MINOR: Remove gradleSetup from Jenkinsfile (#10522)
  MINOR: Use `testRuntimeOnly` instead of `testRuntime` in storage modules (#10524)
  KAFKA-12607; Test case for resigned state vote granting (#10510)
  HOTFIX: delete removed WindowedStore.put() method (#10517)
  KAFKA-8410: Migrate KStream Stateless operators to new Processor API (#10381)
  KAFKA-9831: increase max.poll.interval.ms to avoid unexpected rebalance (#10301)
  MINOR: Fix typo in docs (#10423)
  KAFKA-12449: Remove deprecated WindowStore#put (#10293)
  MINOR: Fix left/outer join descriptions in Streams DSL reference topic (#10453)
  MINOR: add missing TimeoutException to Producer.send JavaDocs (#10495)
  KAFKA-12591; Remove deprecated `quota.producer.default` and `quota.consumer.default` configurations (#10427)
  KAFKA-12492: Fix the formatting of example RocksDBConfigSetter (#10486)
  KAFKA-12406 Integrate client quotas with KRaft broker (#10254)
  KAFKA-12619; Raft leader should expose hw only after committing LeaderChange (#10481)
  KAFKA-12457; Add sentinel ID to metadata topic (#10492)
  HOTFIX: add missing upgrade docs
  KAFKA-12630: Remove deprecated KafkaClientSupplier#getAdminClient in Streams (#10502)
  MINOR: Remove redundant code from ReplicaManagerTest#testFetchMessagesWhenNotFollowerForOnePartition (#10501)
  MINOR: Update Crc32Test#testUpdate method with correct Crc32 (#10406)
  ...
